flink-core/src/main/java/org/apache/flink/types/StringValue.java

```diff
@@ -759,57 +765,151 @@ public static String readString(DataInput in) throws IOException {
+
 			// the length we write is offset by one, because a length of zero indicates a null value
-			int lenToWrite = cs.length()+1;
+			int lenToWrite = strlen + 1;
 			if (lenToWrite < 0) {
 				throw new IllegalArgumentException("CharSequence is too long.");
 			}
-	
-			// write the length, variable-length encoded
-			while (lenToWrite >= HIGH_BIT) {
-				out.write(lenToWrite | HIGH_BIT);
-				lenToWrite >>>= 7;
+
+
+			// buffer size needed to contain all the characters
+			int buflen = 0;
+			if (strlen < 8) {
+				// for small strings it's much faster to over-allocate a buffer sized for the worst-case scenario
+				// when all the chars are encoded as 3-byte sequences, than iterating over the source string.
+				buflen = strlen * 3;
+			} else {
+				for (int i = 0; i < strlen; i++) {
+					char c = cs.charAt(i);
+					if ((c & 0xff80) == 0) {
+						buflen++;
+					} else if (c > 0x07FF) {
+						buflen += 3;
+					} else {
+						buflen += 2;
+					}
+				}
 			}
-			out.write(lenToWrite);
-	
+			byte[] buffer;
+			int position = 0;
+			// string is prefixed by it's variable length encoded size, which can take 1-5 bytes.
+			if (lenToWrite < HIGH_BIT) {
+				buflen += 1;
+				buffer = new byte[buflen];
+				buffer[position++] = (byte) lenToWrite;
+			} else if (lenToWrite < HIGH_BIT14) {
+				buflen += 2;
+				buffer = new byte[buflen];
+				buffer[position++] = (byte)(lenToWrite | HIGH_BIT);
+				buffer[position++] = (byte)((lenToWrite >>> 7));
+			} else if (lenToWrite < HIGH_BIT21) {
+				buflen += 3;
+				buffer = new byte[buflen];
+				buffer[position++] = (byte)(lenToWrite | HIGH_BIT);
+				buffer[position++] = (byte)((lenToWrite >>> 7) | HIGH_BIT);
+				buffer[position++] = (byte)((lenToWrite >>> 14));
+			} else if (lenToWrite < HIGH_BIT28) {
+				buflen += 4;
+				buffer = new byte[buflen];
+				buffer[position++] = (byte)(lenToWrite | HIGH_BIT);
+				buffer[position++] = (byte)((lenToWrite >>> 7) | HIGH_BIT);
+				buffer[position++] = (byte)((lenToWrite >>> 14) | HIGH_BIT);
+				buffer[position++] = (byte)((lenToWrite >>> 21));
+			} else {
+				buflen += 5;
+				buffer = new byte[buflen];
+				buffer[position++] = (byte)(lenToWrite | HIGH_BIT);
+				buffer[position++] = (byte)((lenToWrite >>> 7) | HIGH_BIT);
+				buffer[position++] = (byte)((lenToWrite >>> 14) | HIGH_BIT);
+				buffer[position++] = (byte)((lenToWrite >>> 21) | HIGH_BIT);
+				buffer[position++] = (byte)((lenToWrite >>> 28));
+			}
+
+
 			// write the char data, variable length encoded
-			for (int i = 0; i < cs.length(); i++) {
+			for (int i = 0; i < strlen; i++) {
 				int c = cs.charAt(i);
-	
-				while (c >= HIGH_BIT) {
-					out.write(c | HIGH_BIT);
-					c >>>= 7;
+
+				// manual loop unroll, as it performs much better on jdk8
+				if (c < HIGH_BIT) {
+					buffer[position++] = (byte)c;
+				} else if (c < HIGH_BIT14) {
+					buffer[position++] = (byte)(c | HIGH_BIT);
+					buffer[position++] = (byte)((c >>> 7));
+				} else {
```

Comment:
So I stumbled upon this piece. Since we are encoding a Java `char` as bytes, we will never exceed 3 `bytes`, right?
I think, I would be more assured if we can have a test with a 4 byte unicode. https://www.fileformat.info/info/unicode/char/1f4a3/index.htm
As per my understanding that unicode char would result in two java chars, which are treated individually in this loop, such that they result in 2 bytes each or a total of 4 bytes. Did I get it correctly?

Reply:
Java characters (e.g. char and Character type) have internal UTF16 encoding, meaning that they will always be encoded as 16-bit numbers. But there are some symbols called surrogate pairs, which may require a pair of 16-bit characters to be encoded.

So from the wire format point of view, when you do `"🌉".length()` you will get 2 (as there are two underlying 16-bit characters to encode this symbol), but with `"🌉".codePointCount(0,"🌉".length())` you will get 1 (as there only one single symbol). But the string serializer is not operating on symbols and UTF code points, it is operating on the underlying 16-bit numbers, so it must handle it properly out of the box.

Reply:
@AHeise I've added a trivial test for surrogate pairs serialization, validating that these cases are still handled properly.